### PR TITLE
feat(mls): support calling in MLS groups FS-675

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-675" title="FS-675" target="_blank">TaskFS-675</a>  [iOS] Distribute CONFKEY to all members of conversation [temp-M4]
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR adds support for calling in MLS groups. Changes include:

- process `conversationMLSMessageAdd` event types for incoming calling messages sent via mls
- sending outgoing calling messages via mls
- send CONFKEY to mls group (temp workaround until subgroups are implemented)
- dropping all targeted message except CONFKEY (temp workaround until subgroups are implemented)
- fetching clients list without the need of sending an OTR message

### Dependencies

Needs releases with:

- https://github.com/wireapp/wire-ios-request-strategy/pull/561

#### Test Coverage

- unit tests for sending calling messages
- unit tests for fetching client list

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Cartfile.resolved [hunk 1]
```diff
@@ -15,7 +15,7 @@ github "wireapp/wire-ios-protos" "32.0.0"
 github "wireapp/wire-ios-system" "38.0.0"
 github "wireapp/wire-ios-testing" "29.0.0"
 github "wireapp/wire-ios-utilities" "47.0.0"
-github "wireapp/wire-ios-data-model" "feat/mls"
+github "wireapp/wire-ios-data-model" "fix/mls-event-type-checks"
```
Comment (johnxnguyen) [inline Cartfile.resolved:18]:
```suggestion
github "wireapp/wire-ios-data-model" "fix/mls"
```

File: Cartfile.resolved [hunk 2]
```diff
@@ -15,7 +15,7 @@ github "wireapp/wire-ios-protos" "32.0.0"
 github "wireapp/wire-ios-system" "38.0.0"
 github "wireapp/wire-ios-testing" "29.0.0"
 github "wireapp/wire-ios-utilities" "47.0.0"
-github "wireapp/wire-ios-data-model" "feat/mls"
+github "wireapp/wire-ios-data-model" "fix/mls"
```
Comment (johnxnguyen) [inline Cartfile.resolved:18]:
```suggestion
github "wireapp/wire-ios-data-model" "feat/mls"
```

File: Source/Synchronization/Strategies/CallingRequestStrategy.swift [hunk 3]
```diff
@@ -251,7 +253,7 @@ public final class CallingRequestStrategy: AbstractRequestStrategy, ZMSingleRequ
             processCallEvent(
                 conversationUUID: conversationUUID,
                 senderUUID: senderUUID,
-                clientId: clientId,
+                clientId: event.senderClientID ?? callEventContent.callerClientID,
```
Comment (johnxnguyen) [inline Source/Synchronization/Strategies/CallingRequestStrategy.swift:256]: the sender client id doesn't exist in mls update events, so we take it from the call message instead.

File: Tests/Source/Synchronization/Strategies/CallingRequestStrategyTests.swift [hunk 4]
```diff
@@ -448,3 +795,56 @@ class CallingRequestStrategyTests: MessagingTest {
     }
 
 }
+
+class MockMLSController: MLSControllerProtocol {
```
Comment (johnxnguyen) [inline Tests/Source/Synchronization/Strategies/CallingRequestStrategyTests.swift:799]: It's annoying that we have to create yet another mock for this protocol... let's consider it tech debt for now.

Comment (David-Henner) [reply Tests/Source/Synchronization/Strategies/CallingRequestStrategyTests.swift:799]: maybe a quick & dirty fix would be to change the target of the `MLSControllerProtocol` mock of data model and make it public

File: Source/Synchronization/Strategies/CallingRequestStrategy.swift [hunk 5]
```diff
@@ -531,3 +591,52 @@ extension CallingRequestStrategy {
...
+
+    public func fetchUserClients(
+        userIDs: Set<QualifiedID>,
+        in context: NSManagedObjectContext
+    ) async throws -> Set<QualifiedClientID> {
+        var action = FetchUserClientsAction(userIDs: userIDs)
+        return try await action.perform(in: context.notificationContext)
+    }
+
+}
```
Comment (David-Henner) [inline Source/Synchronization/Strategies/CallingRequestStrategy.swift:642]: we could extract this to another file, what do you think?

Comment (johnxnguyen) [reply Source/Synchronization/Strategies/CallingRequestStrategy.swift]: yep, done

File: Tests/Source/Synchronization/Strategies/CallingRequestStrategyTests.swift (hunk 4 above)
Comment (johnxnguyen) [reply Tests/Source/Synchronization/Strategies/CallingRequestStrategyTests.swift:799]: Yeah we can do that, but I'll do it in a separate PR.

File: Cartfile.resolved [hunk 6]
```diff
@@ -17,5 +17,5 @@ github "wireapp/wire-ios-testing" "29.0.0"
 github "wireapp/wire-ios-utilities" "47.0.0"
 github "wireapp/wire-ios-data-model" "feat/mls"
 github "wireapp/wire-ios-mocktransport" "feat/mls"
-github "wireapp/wire-ios-request-strategy" "feat/mls"
+github "wireapp/wire-ios-request-strategy" "feat/mls-calling"
```
Comment (johnxnguyen) [inline Cartfile.resolved:20]:
```suggestion
github "wireapp/wire-ios-request-strategy" "feat/mls"
```

**Pull request merged**